### PR TITLE
Convert Notebook tests to papermill

### PR DESCRIPTION
**Pull request by riedgar-ms**
Rather than using `nbval`, convert our notebook tests to use `papermill`. With the help of `nteract-scrapbook` we can then examine the contents of particular variables from the notebooks to ensure that we're getting the expected results.

Explicit `scrapbook` commands are required to save out values for future examination, but we don't want to include these when our users look at the notebooks. Accordingly, we include machinery for adding the necessary cells to the notebooks dynamically.

### CODE REVIEW

**Comment by riedgar-ms** [comment]
I'll wait for comments before trying to convert the rest of the notebooks

File: devops/notebook-job-template.yml [hunk 1]
```diff
@@ -35,7 +35,7 @@ jobs:
   - script: pip install -e .
     displayName: 'Install fairlearn in edit mode'
 
-  - script: python -m pytest notebooks/ --nbval --junitxml=./TEST-TEST.xml -o junit_suite_name="$(Agent.JobName)"
+  - script: python -m pytest test/notebooks/ --junitxml=./TEST-TEST.xml -o junit_suite_name="$(Agent.JobName)"
```
Comment (riedgar-ms) [inline devops/notebook-job-template.yml]: Also need to change the templates used in checking the PyPI releases

File: devops/all-tests-job-template.yml [hunk 2]
```diff
@@ -44,7 +44,7 @@ jobs:
   - script: flake8 .
     displayName: "Run flake8"
 
-  - script: python -m pytest test/ --ignore=test/perf --junitxml=./TEST--TEST.xml -o junit_suite_name="$(Agent.JobName)"
+  - script: python -m pytest test/unit --junitxml=./TEST--TEST.xml -o junit_suite_name="$(Agent.JobName)"
```
Comment (romanlutz) [inline devops/all-tests-job-template.yml]: why this change? Explicitly ignoring means we'd catch issues in new tests that are in neither directory, while this now would ignore test/integration or whatever one might add

Comment (riedgar-ms) [reply devops/all-tests-job-template.yml]: Have changed, but we're getting to the point where these need to be regularised some more.

File: devops/notebook-job-template.yml (hunk 1 above)
Comment (riedgar-ms) [reply devops/notebook-job-template.yml]: Done

**Comment by bethz** [comment]
looks good!
